### PR TITLE
Adds Query Datum Support for a Form

### PR DESCRIPTION
**Pull request by shubham1g5**
https://dimagi-dev.atlassian.net/browse/USH-1231

Preserves the Session Frame Inside the `FormSession` to load up any query xml data instances into the Form Session

cross-request: https://github.com/dimagi/commcare-core/pull/1040

### CODE REVIEW

**Review by snopoke — COMMENTED** [review]
Code looks OK to me.

File: src/test/java/org/commcare/formplayer/tests/FormEntryWithQueryTests.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        // see if the instance is retained into the form session
+        EvaluateXPathResponseBean evaluateXPathResponseBean = evaluateXPath(responseBean.getSessionId(),
+                "instance('registry')/results/case[@case_type='case']/case_name");
+        assert evaluateXPathResponseBean.getStatus().equals(Constants.ANSWER_RESPONSE_STATUS_POSITIVE);
+        String result = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<result>Burt Maclin</result>\n";
+        assert evaluateXPathResponseBean.getOutput().equals(result);
+    }
+
+    @Test
+    public void testCaseSearchWithRegistryNavigation() throws Exception {
```
Comment (shubham1g5) [inline src/test/java/org/commcare/formplayer/tests/FormEntryWithQueryTests.java:70]: this fails right now due to menu navigation breaking for case selection from a registry

File: src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java [hunk 2]
```diff
@@ -36,27 +37,31 @@
 
     private final Log log = LogFactory.getLog(CaseSearchHelper.class);
 
-    public ExternalDataInstance getSearchDataInstance(FormplayerQueryScreen screen, URI uri) {
+    @Override
+    public ExternalDataInstance getRemoteDataInstance(String instanceId, boolean useCaseTemplate, URI uri)
+            throws UnfullfilledRequirementsException, XmlPullParserException, InvalidStructureException, IOException {
         Cache cache = cacheManager.getCache("case_search");
         String cacheKey = getCacheKey(uri);
         TreeElement cachedRoot = cache.get(cacheKey, TreeElement.class);
```
Comment (snopoke) [inline src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java:45]: not really related to this PR but how does this cache get cleared? When I was testing locally I wasn't able to get updated case list results because of this cache.

File: src/main/java/org/commcare/formplayer/services/MenuSessionRunnerService.java [hunk 3]
```diff
@@ -403,27 +414,39 @@ private NotificationMessage doQuery(FormplayerQueryScreen screen,
...
+            searchDataInstance = searchAndSetResult(
+                    screen,
+                    screen.getUri(skipDefaultPromptValues));
+            if (searchDataInstance == null) {
+                error = "No result from query";
             }
-        } else {
+        } catch (InvalidStructureException | IOException
+                | XmlPullParserException | UnfullfilledRequirementsException e) {
+            e.printStackTrace();
```
Comment (snopoke) [inline src/main/java/org/commcare/formplayer/services/MenuSessionRunnerService.java:428]: nit: printStackTrace

File: src/main/java/org/commcare/formplayer/session/FormSession.java [hunk 4]
```diff
@@ -505,8 +550,8 @@ public FormEntryNavigationResponseBean getFormNavigation() throws IOException {
     }
 
     /**
-     *  Automatically advance to the next question after an answer when in "prompt" mode
-     *  (currently only used by SMS)
```
Comment (snopoke) [inline src/main/java/org/commcare/formplayer/session/FormSession.java:509]: nit: all these formatting changes make it harder to see what actually changed in the PR - perhaps put them in a separate PR? (also maybe turn off auto-formatting in your IDE?)

File: src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java [hunk 5]
```diff
@@ -36,27 +37,31 @@
...
+                    instanceId,
+                    copyOfRoot,
+                    uri.toString(),
+                    useCaseTemplate);
         }
-        log.info(String.format("Making case search request to url %s",  uri));
+        log.info(String.format("Making case search request to url %s", uri));
         String responseString = webClient.get(uri);
         if (responseString != null) {
-            Pair<ExternalDataInstance, String> dataInstanceWithError = screen.processResponse(
```
Comment (snopoke) [inline src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java:51]: is `screen.processResponse(` still called anywhere or can it be removed?

File: src/main/java/org/commcare/formplayer/session/FormplayerInstanceInitializer.java [hunk 6]
```diff
@@ -60,11 +60,6 @@ protected AbstractTreeElement setupSessionData(ExternalDataInstance instance) {
         if (u == null) {
             throw new RuntimeException("There was a problem loading the user data. Please Sync.");
         }
-        if (sessionData != null) {
-            for (String key : sessionData.keySet()) {
-                session.setDatum(key, sessionData.get(key));
-            }
-        }
```
Comment (snopoke) [inline src/main/java/org/commcare/formplayer/session/FormplayerInstanceInitializer.java:67]: is this logic replaced by `createSessionFrame`

File: src/main/java/org/commcare/formplayer/session/FormSession.java (hunk 4 above)
Comment (shubham1g5) [reply src/main/java/org/commcare/formplayer/session/FormSession.java:509]: Yeah this is definitely a bit of an issue. Ideally, I would like to retain the auto-format capability if possible. Do you think there is a possibility here to abide to a [style file like mobile](https://github.com/dimagi/commcare-android#code-style-settings) and do a PR that does a one time format with that style file. This would avoid lot of these diffs in future though would require us to use an IntelliJ based IDE to make these edits.

Comment (snopoke) [reply src/main/java/org/commcare/formplayer/session/FormSession.java:509]: seems reasonable, do you know if there is a way to integrate that into the PR process? Can we do a style check?

**Comment by orangejenny** [comment]
Closing in favor of https://github.com/dimagi/formplayer/pull/1028
